### PR TITLE
[RDY] Problem analysis

### PR DESCRIPTION
**Pull request by gfokkema**

### CODE REVIEW

File: research/05-problem-ana.tex [hunk 1]
```diff
@@ -1,10 +1,134 @@
...
+This section lists the problems that are expected when trying to
+implement generating REPLs generically. For each problem, their
+solution approaches that should be explored during the implementation
+process are given, as well as the parts of Spoofax that are relevant
+for that problem.
+
+\subsubsection{Detecting unfinished expressions for multiline editing}
+\label{sec:detect-unfin-expr}
+
+\subsubsection{Language specific additional commands}
```
Comment (Hjdskes) [inline research/05-problem-ana.tex:33]: What requirements does this refer to?

File: research/05-problem-ana.tex [hunk 2]
```diff
@@ -1,10 +1,134 @@
...
+Spoofax: when one creates a language, there is the possibility to
+define menu buttons and bind them to Stratego strategies
+(\cref{sec:orgheadline5}). For example, one can define a menu button
+``Run'' to run the expression that is currently selected (see
+\cref{fig:menu-actions}), or a ``Desugar'' button to desugar the
+program that is currently open. In the same way, one may define a
+``list-modules'' command for a REPL which is bound to a Stratego
+strategy.
+
+\subsubsection{Redefining terms bound to names}
```
Comment (Unia) [inline research/05-problem-ana.tex]: Isn't this a subsubsubsection?

File: research/05-problem-ana.tex [hunk 3]
```diff
@@ -1,10 +1,134 @@
...
+\begin{itemize}
+\item a special prefix for shell specific commands, and then...
+\item a command to enumerate all possible namespaces as detected by spoofax
+\item a command to switch to a namespace from this list
+\item a command to list entries in the current namespace
+\item expressions typed will be added to the selected namespace
+\item all language agnostic, since we won't care what to call the namespace
+\end{itemize}
+
+\subsubsection{Partially replacing the AST of the current program}
```
Comment (Hjdskes) [inline research/05-problem-ana.tex:94]: Isn't this a subsubsubsection?

File: research/05-problem-ana.tex (hunk 1 above)
Comment (Balletie) [inline research/05-problem-ana.tex:33]: What do you mean

File: research/05-problem-ana.tex [hunk 4]
```diff
@@ -1,10 +1,127 @@
 \section{Problem Analysis}
 \label{sec:problem-analysis}
+This section gives an analysis of the problem as defined in the
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: I think the sentence is better when you remove `as`.

File: research/05-problem-ana.tex [hunk 5]
```diff
@@ -1,10 +1,127 @@
...
+previous section \ref{sec:problem-definition}. More specifically, this
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: The reference can be removed imo.

File: research/05-problem-ana.tex [hunk 6]
```diff
@@ -1,10 +1,127 @@
...
+section describes the problems that are expected and to be solved
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: `and have to be`

File: research/05-problem-ana.tex [hunk 7]
```diff
@@ -1,10 +1,127 @@
...
+during the implementation process of the product. The first part of
+this section describes the process of incorporating REPL generation
+within the larger context of Spoofax. The second part analyzes the
+part of the problem concerned with literate programming, but this part
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: `programming. This part...`

File: research/05-problem-ana.tex [hunk 8]
```diff
@@ -1,10 +1,127 @@
...
+will be analyzed in less detail than generating REPLs as it is unsure
+whether this will be present in the final product.
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: `this` -> `it`?

File: research/05-problem-ana.tex [hunk 9]
```diff
@@ -1,10 +1,127 @@
...
 
-\subsection{How does REPL generation fit within Spoofax}
-\label{ssec:architecture}
+\subsection{Incorporating REPL generation within Spoofax}
+\label{ssec:incorp-repl-gener}
+To generate solidly working REPLs for any language defined in Spoofax,
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: `solidly` looks a bit strange. Is `properly` better?

File: research/05-problem-ana.tex [hunk 10]
```diff
@@ -1,10 +1,127 @@
...
+some Spoofax concepts have to be carefully considered. Since every
+language can have many different language constructs, generating REPLs
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: This is a very long sentence. I suggest the following:

> Every language can have many different language constructs. Generating REPLs, however, should not be constrained to any particular language constructs such as classes. Rather, the generation should be done generically. For this reason the implementation and particularly the interaction with the user needs to be carefully considered.

File: research/05-problem-ana.tex [hunk 11]
```diff
@@ -1,10 +1,127 @@
...
+should not be constrained to language specific constructs like
+classes, but should rather be done generically. Therefore the
+implementation and particularly the interaction with the user needs to
+be thought of carefully.
 
-\subsection{How are plugins developed for Eclipse?}
+This section lists the problems that are expected when trying to
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: You already said this in the introduction to this section.

File: research/05-problem-ana.tex [hunk 12]
```diff
@@ -1,10 +1,127 @@
...
+implement generating REPLs generically. For each problem, their
+solution approaches that should be explored during the implementation
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: This sentence is a bit strange; I'm not sure what you're trying to say here...

File: research/05-problem-ana.tex [hunk 13]
```diff
@@ -1,10 +1,127 @@
...
+process are given, as well as the parts of Spoofax that are relevant
+for that problem.
+
+\subsubsection{Detecting unfinished expressions for multiline editing}
+\label{sec:detect-unfin-expr}
+To support multiline editing, the generated REPL should detect that
+the expression is unfinished when the user presses ``enter'' for the
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: > the ``Return'' key

?

File: research/05-problem-ana.tex [hunk 14]
```diff
@@ -1,10 +1,127 @@
...
+next line, instead of trying to parse and execute the unfinished
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: `the` -> `an`. Also, there are statements and expressions -- at least, I have made that distinction in my parts. Perhaps you should use `process an unfinished input` instead.

File: research/05-problem-ana.tex [hunk 15]
```diff
@@ -1,10 +1,127 @@
...
+expression. An obvious part of Spoofax that is relevant for this
+problem, is the syntax definition in SDF3 (\cref{sec:orgheadline1}).
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: `(see \cref{sec:orgheadline1})`?

File: research/05-problem-ana.tex [hunk 16]
```diff
@@ -1,10 +1,127 @@
...
+
+\subsubsection{Language specific additional commands}
```
Comment (Hjdskes) [inline research/05-problem-ana.tex:33]: The feature that is listed in section 2 and section 4 is that of REPL specific commands, not necessarily in the language's syntax. This seems to be a different problem, or at least it is formulated differently.

File: research/05-problem-ana.tex [hunk 17]
```diff
@@ -1,10 +1,127 @@
...
+\label{sec:lang-spec-addit}
+Another problem is when some additional command can be useful for a
+REPL of a particular language, but would not make any sense within the
+context of other languages. For example, some languages such as Java
+have the notion of a ``module'' or ``package'', which declares a
+separate namespace for the definitions in it. An additional command to
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: `in` -> `inside`.

File: research/05-problem-ana.tex [hunk 18]
```diff
@@ -1,10 +1,127 @@
...
+\label{sec:detect-unfin-expr}
+To support multiline editing, the generated REPL should detect that
+the expression is unfinished when the user presses ``enter'' for the
+next line, instead of trying to parse and execute the unfinished
+expression. An obvious part of Spoofax that is relevant for this
+problem, is the syntax definition in SDF3 (\cref{sec:orgheadline1}).
+
+\subsubsection{Language specific additional commands}
+\label{sec:lang-spec-addit}
+Another problem is when some additional command can be useful for a
```
Comment (Hjdskes) [inline research/05-problem-ana.tex:35]: > Another problem is that of additional commands that can be useful for a REPL of a particular language, whilst not making sense within the context of other languages.

File: research/05-problem-ana.tex [hunk 19]
```diff
@@ -1,10 +1,127 @@
...
+REPL of a particular language, but would not make any sense within the
+context of other languages. For example, some languages such as Java
+have the notion of a ``module'' or ``package'', which declares a
+separate namespace for the definitions in it. An additional command to
+list all of the currently loaded modules is useful, but would not make
+any sense in languages that have no concept of a module.
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: Aren't all these things defined in their separate namespaces, whatever there may be? **All** languages in Spoofax use NaBL for their name bindings, so **all** languages in Spoofax have these namespaces that we can list. As such, I think the example issue given here does not make much sense. Even more so, because **all** languages have these namespaces, we can just add a language-agnostic command enabled in **all** REPLs to list these namespaces. That is precisely what the feature "Additional commands to the REPL" means.

File: research/05-problem-ana.tex [hunk 20]
```diff
@@ -1,10 +1,127 @@
...
+
+This problem could be solved by the language designer by extending his
+or her language with reflective capabilities. However in that case it
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: `his or her` -> `their`.

File: research/05-problem-ana.tex [hunk 21]
```diff
@@ -1,10 +1,127 @@
...
+would not be an additional command, but just an expression like any
+other expression in that language. Moreover, the language designer
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: What's the problem with this?

File: research/05-problem-ana.tex [hunk 22]
```diff
@@ -1,10 +1,127 @@
...
+  \centering
+  \includegraphics[width=0.5\textwidth]{menu-actions.png}
+  \caption{A menu action for the paplj language defined using
+    Spoofax. The bottom window shows the menu definition, the top
+    window shows a program written in paplj.}
+  \label{fig:menu-actions}
+\end{figure}
+
+One possible solution is to allow for language specific configurations
+that are loaded and used during the generation process of a REPL for
```
Comment (Hjdskes) [inline research/05-problem-ana.tex:60]: Aren't they loaded whenever the REPL is started, and not during the generation process?

File: research/05-problem-ana.tex [hunk 23]
```diff
@@ -1,10 +1,127 @@
...
+that language. This is similar to something that currently exists in
+Spoofax: when one creates a language, there is the possibility to
+define menu buttons and bind them to Stratego strategies
+(\cref{sec:orgheadline5}). For example, one can define a menu button
+``Run'' to run the expression that is currently selected (see
+\cref{fig:menu-actions}), or a ``Desugar'' button to desugar the
+program that is currently open. In the same way, one may define a
+``list-modules'' command for a REPL which is bound to a Stratego
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: `modules` -> `namespaces`?

File: research/05-problem-ana.tex [hunk 24]
```diff
@@ -1,10 +1,127 @@
...
+strategy.
+
+\subsubsection{Redefining terms bound to names}
+\label{sec:redef-cont-bound}
+When the user is prototyping methods inside the REPL, he or she would
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: `he or she` -> `they`.

File: research/05-problem-ana.tex [hunk 25]
```diff
@@ -1,10 +1,127 @@
...
+probably want to be able to redefine that method to be of a different
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: `Probably` is the wrong word here. I would rephrase this as: `it is not unlikely that they want to be able to ...`

File: research/05-problem-ana.tex [hunk 26]
```diff
@@ -1,10 +1,127 @@
...
+implementation. However, this poses a similar problem as in the
+previous section, because for some languages it may not be possible
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: `section: for some languages..`.

File: research/05-problem-ana.tex [hunk 27]
```diff
@@ -1,10 +1,127 @@
...
+nor desirable to do so outside of the context of a REPL. Thus
+requiring the language designer to extend the language with such
+abilities is again an inadequate solution.
+
+One could propose the same solution as in the previous section, namely
+to define an additional command for redefining a class or
+method. However, in this case this is not an adequate solution either,
+since binding terms to names is a common feature among programming
```
Comment (Hjdskes) [inline research/05-problem-ana.tex:84]: What makes it not an adequate solution then?

File: research/05-problem-ana.tex [hunk 28]
```diff
@@ -1,10 +1,127 @@
...
+languages. The fact that Spoofax comes with the NaBL DSL for
+specifying name binding rules illustrates this. Name binding is in
+fact the very feature which enables one to create abstractions.
```
Comment (Hjdskes) [inline research/05-problem-ana.tex:87]: Which abstractions?

File: research/05-problem-ana.tex [hunk 29]
```diff
@@ -1,10 +1,127 @@
...
+
+A possible and more adequate solution for redefining a term bound to a
+name, is to allow the user to give the name and the new term. The name
+can then be used to find the old term, so that it can be replaced with
+the new one.
+
+\subsubsection{Partially replacing the AST of the current program}
+When given an identifier to bind a new term to, the leaves in the AST
+corresponding to the identifier have to be replaced by the new terms.
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: `replaced by` or `replaced with`?

File: research/05-problem-ana.tex [hunk 30]
```diff
@@ -1,10 +1,127 @@
...
+This implies that the REPL needs to able parse single terms that might
+not be valid programs, and then update the AST of the program accordingly.
+This requirement has certain effects when it comes to the build process
+used in the REPL.
+
+First of all, the REPL should always hold on to the AST representation of the
+executing program, in order to find the old term in the current program.
```
Comment (Hjdskes) [inline research/05-problem-ana.tex:103]: So executing == current. Perhaps you can change this to `such that it is able to find the old term`.

File: research/05-problem-ana.tex [hunk 31]
```diff
@@ -1,10 +1,127 @@
...
+Since Spoofax can just as well transform the AST back to original syntax,
+it is possible to alter this identifier either directly as a Stratego term or
+in the original syntax.
+
+Secondly the build process in Eclipse uses a `BuildInputBuilder' to transform
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: Comma after secondly.

Comment (Hjdskes) [inline research/05-problem-ana.tex]: `\texttt{BuildInputBuilder}`?

File: research/05-problem-ana.tex [hunk 32]
```diff
@@ -1,10 +1,127 @@
...
+and build a (valid) program as a whole. While this is quite a convenient way to
+quickly create a compiled program, it is desirable that the REPL has a bigger
+influence on the way programs are built.
+Therefore the REPL should be able to execute individual Spoofax stages
```
Comment (Hjdskes) [inline research/05-problem-ana.tex]: > The REPL should therefore be able to ...

File: research/05-problem-ana.tex (hunk 5 above)
Comment (Balletie) [inline research/05-problem-ana.tex]: ...

Comment (Hjdskes) [inline research/05-problem-ana.tex]: Don't you think it sounds weird? `In the previous section 4`.

File: research/05-problem-ana.tex (hunk 16 above)
Comment (Balletie) [inline research/05-problem-ana.tex:33]: This is talking about exactly the same thing, commands for the REPL that make sense only for a specific language.

File: research/05-problem-ana.tex (hunk 19 above)
Comment (Balletie) [inline research/05-problem-ana.tex]: You're not forced to use NaBL or TS.

File: research/05-problem-ana.tex (hunk 21 above)
Comment (Balletie) [inline research/05-problem-ana.tex]: Ok I will remove it

File: research/05-problem-ana.tex (hunk 16 above)
Comment (Hjdskes) [inline research/05-problem-ana.tex:33]: Sorry, you are correct.

File: research/05-problem-ana.tex (hunk 23 above)
Comment (Balletie) [inline research/05-problem-ana.tex]: The example I gave above talked about modules or packages.

File: research/05-problem-ana.tex (hunk 19 above)
Comment (Hjdskes) [inline research/05-problem-ana.tex]: But then we can also argue that you don't have to use DynSem, or Stratego, or ... Being generic and agnostic has to stop at some point or we're left with no starting point whatsoever. I think assuming the usage of NabL or TS is a reasonable assumption to make.

File: research/05-problem-ana.tex (hunk 28 above)
Comment (Balletie) [inline research/05-problem-ana.tex:87]: Functions, for example. If you can't name things, you can't create abstractions. You need to be able to assign a name to a "thing" in order to abstract.

Even the simple lambda calculus allows for abstractions via parameter names and function applications. For that you need name binding. Without something similar to name binding and function application, how are you going to create an abstraction?

File: research/05-problem-ana.tex (hunk 31 above)
Comment (Hjdskes) [inline research/05-problem-ana.tex]: Otherwise, use `` ... ''. I'm applying this consistently everywhere atm.

File: research/05-problem-ana.tex (hunk 27 above)
Comment (Balletie) [inline research/05-problem-ana.tex:84]: It's a feature that is common to all programming languages, as opposed to language specific constructs like in the previous section. Therefore using the same solution again is not adequate, as it is not entirely the same problem. It requires a solution more tuned to this problem.

File: research/05-problem-ana.tex (hunk 22 above)
Comment (Balletie) [inline research/05-problem-ana.tex:60]: No I don't think so, because it is a specific feature of the REPL right? Don't think it really matters though, it's more about the ability to configure.

File: research/05-problem-ana.tex (hunk 19 above)
Comment (Balletie) [inline research/05-problem-ana.tex]: Anyway, paplj for example has no notion of a module.

File: research/05-problem-ana.tex (hunk 5 above)
Comment (Balletie) [inline research/05-problem-ana.tex]: yea it's just that previously you wanted me to add references to sections in these cases, so it's what I did here :P. Should I remove the references to previous/next section also in the spoofax section then?

File: research/05-problem-ana.tex (hunk 22 above)
Comment (Hjdskes) [inline research/05-problem-ana.tex:60]: What I mean is that if you are to switch to a different language, the items added to the menus by the IDE services of the current language should be removed.

File: research/05-problem-ana.tex (hunk 10 above)
Comment (Balletie) [inline research/05-problem-ana.tex]: Ok, I split it but a little different

File: research/05-problem-ana.tex (hunk 11 above)
Comment (Balletie) [inline research/05-problem-ana.tex]: Not entirely, here it's specifically for REPLs. But I will try to remove some duplication.

File: research/05-problem-ana.tex (hunk 14 above)
Comment (Balletie) [inline research/05-problem-ana.tex]: well, the _expression_ is unfinished, that's the whole point. The expression is not done yet, because it's not syntactically valid.

Comment (Hjdskes) [inline research/05-problem-ana.tex]: But _statements_ can also be unfinished.

File: research/05-problem-ana.tex (hunk 18 above)
Comment (Balletie) [inline research/05-problem-ana.tex:35]: Don't really think this is an improvement

File: research/05-problem-ana.tex (hunk 14 above)
Comment (Balletie) [inline research/05-problem-ana.tex]: Is it really that crucial?

Comment (Hjdskes) [inline research/05-problem-ana.tex]: I'm just saying that in my section I have made the distinction. It's up to you if you make it here as well, but it's only a small change...

File: research/05-problem-ana.tex (hunk 22 above)
Comment (Balletie) [inline research/05-problem-ana.tex:60]: Yea but that's different I think. A REPL is not like a menu that it should adapt, depending on what window is currently active.

Comment (Hjdskes) [inline research/05-problem-ana.tex:60]: But isn't the whole suggestion here to add entries into the menus?

Comment (Balletie) [inline research/05-problem-ana.tex:60]: No, the suggestion is to allow for defining actions bound to REPL command, _similar to_ how actions can currently be bound to menu entries.

File: research/05-problem-ana.tex (hunk 14 above)
Comment (Balletie) [inline research/05-problem-ana.tex]: ok
